File: web/lib/api/accounts.ts
import { apiFetch } from "./client"
import type {
  AccountStatusResponse,
  JobOut,
  LoginRequest,
  StoreCookieRequest,
  StoreCredentialsRequest,
} from "./types"

export function getAccountStatus(
  platform: string
): Promise<AccountStatusResponse> {
  return apiFetch<AccountStatusResponse>(
    `/api/v1/accounts/${encodeURIComponent(platform)}/status`
  )
}

/** 憑證只往後端加密 vault 送，送出後前端立即清空 state、永不落 localStorage。 */
export function storeCredentials(
  platform: string,
  body: StoreCredentialsRequest
): Promise<void> {
  return apiFetch<void>(
    `/api/v1/accounts/${encodeURIComponent(platform)}/credentials`,
    { method: "PUT", body }
  )
}

export function storeCookie(
  platform: string,
  body: StoreCookieRequest
): Promise<void> {
  return apiFetch<void>(
    `/api/v1/accounts/${encodeURIComponent(platform)}/cookie`,
    { method: "POST", body }
  )
}

export function eraseCredentials(platform: string): Promise<void> {
  return apiFetch<void>(
    `/api/v1/accounts/${encodeURIComponent(platform)}/credentials`,
    { method: "DELETE" }
  )
}

/** profile 是 query 參數，不是 body（src/api/routers/accounts.py:75）。 */
export function requestSessionCheck(
  platform: string,
  profile = "live"
): Promise<JobOut> {
  return apiFetch<JobOut>(
    `/api/v1/accounts/${encodeURIComponent(platform)}/session/check`,
    { method: "POST", query: { profile } }
  )
}

export function requestLogin(
  platform: string,
  body: LoginRequest
): Promise<JobOut> {
  return apiFetch<JobOut>(
    `/api/v1/accounts/${encodeURIComponent(platform)}/login`,
    {
      method: "POST",
      body,
    }
  )
}

export function getAccountJob(jobId: string): Promise<JobOut> {
  return apiFetch<JobOut>(`/api/v1/accounts/jobs/${encodeURIComponent(jobId)}`)
}

const TERMINAL_JOB_STATES = ["DONE", "FAILED", "CANCELLED"]

export function isTerminalJobState(state: string): boolean {
  return TERMINAL_JOB_STATES.includes(state)
}

File: src/api/routers/accounts.py
from __future__ import annotations

from fastapi import APIRouter, Depends, Response, status

from accounts.service import AccountService, VaultNotConfiguredError
from broker.broker import SqliteTaskBroker

from ..deps import get_accounts, get_broker
from ..errors import InvalidRequestError, NotFoundError, VaultLockedError
from ..schemas.accounts import (
    AccountStatusResponse,
    JobOut,
    LoginRequest,
    StoreCookieRequest,
    StoreCredentialsRequest,
)

router = APIRouter(prefix="/api/v1/accounts", tags=["accounts"])


@router.get("/status", response_model=AccountStatusResponse)
async def get_default_account_status(
    accounts: AccountService = Depends(get_accounts),
) -> AccountStatusResponse:
    info = accounts.status("kktix")
    return AccountStatusResponse(
        platform=info.platform,
        source=info.source,
        configured=info.configured,
        masked_account=info.masked_account,
        credential_kind=info.credential_kind.value if info.credential_kind else None,
    )


@router.get("/{platform}/status", response_model=AccountStatusResponse)
async def get_platform_account_status(
    platform: str,
    accounts: AccountService = Depends(get_accounts),
) -> AccountStatusResponse:
    info = accounts.status(platform)
    return AccountStatusResponse(
        platform=info.platform,
        source=info.source,
        configured=info.configured,
        masked_account=info.masked_account,
        credential_kind=info.credential_kind.value if info.credential_kind else None,
    )


@router.put("/{platform}/credentials", status_code=status.HTTP_204_NO_CONTENT)
async def store_platform_credentials(
    platform: str,
    req: StoreCredentialsRequest,
    accounts: AccountService = Depends(get_accounts),
) -> Response:
    try:
        accounts.store(platform, req.account, req.access_key)
    except VaultNotConfiguredError as exc:
        raise VaultLockedError(str(exc)) from exc
    return Response(status_code=status.HTTP_204_NO_CONTENT)


@router.post("/{platform}/cookie", status_code=status.HTTP_204_NO_CONTENT)
@router.put("/{platform}/cookie", status_code=status.HTTP_204_NO_CONTENT)
async def store_platform_cookies(
    platform: str,
    req: StoreCookieRequest,
    accounts: AccountService = Depends(get_accounts),
) -> Response:
    try:
        accounts.store_cookie(platform, req.cookies)
    except VaultNotConfiguredError as exc:
        raise VaultLockedError(str(exc)) from exc
    except ValueError as exc:
        raise InvalidRequestError(str(exc)) from exc
    return Response(status_code=status.HTTP_204_NO_CONTENT)


@router.delete("/{platform}/credentials", status_code=status.HTTP_204_NO_CONTENT)
async def erase_platform_credentials(
    platform: str,
    accounts: AccountService = Depends(get_accounts),
) -> Response:
    accounts.erase(platform)
    return Response(status_code=status.HTTP_204_NO_CONTENT)


@router.post(
    "/{platform}/session/check",
    response_model=JobOut,
    status_code=status.HTTP_202_ACCEPTED,
)
async def request_session_check(
    platform: str,
    profile: str = "live",
    accounts: AccountService = Depends(get_accounts),
) -> JobOut:
    job_id = await accounts.request_session_check(platform, profile=profile)
    return JobOut(
        job_id=job_id,
        kind="SESSION_CHECK",
        state="PENDING",
    )


@router.post(
    "/{platform}/login",
    response_model=JobOut,
    status_code=status.HTTP_202_ACCEPTED,
)
async def request_login(
    platform: str,
    req: LoginRequest,
    accounts: AccountService = Depends(get_accounts),
) -> JobOut:
    job_id = await accounts.request_login(platform, profile=req.profile, mode=req.mode)
    kind_str = "AUTO_LOGIN" if req.mode == "auto" else "MANUAL_LOGIN"
    return JobOut(
        job_id=job_id,
        kind=kind_str,
        state="PENDING",
    )


@router.get("/jobs/{job_id}", response_model=JobOut)
async def get_account_job(
    job_id: str,
    broker: SqliteTaskBroker = Depends(get_broker),
) -> JobOut:
    job = await broker.get(job_id)
    if job is None:
        raise NotFoundError(f"Job {job_id} not found")
    return JobOut(
        job_id=job.id,
        kind=job.kind.value,
        state=job.state.value,
        result=job.result,
        error=job.error,
    )
